d2/src/main/java/com/linkedin/d2/balancer/clients/RetryClient.java

```diff
@@ -254,16 +350,134 @@ public void onError(Throwable e)
 
     private boolean isRetryException(Throwable e)
     {
-      return ExceptionUtils.indexOfType(e, RetriableRequestException.class) != -1;
+      Throwable[] throwables = ExceptionUtils.getThrowables(e);
+
+      for (Throwable throwable: throwables)
+      {
+        if (throwable instanceof RetriableRequestException)
+        {
+          return !((RetriableRequestException) throwable).getDoNotRetryOverride();
+        }
+      }
+
+      return false;
     }
 
     /**
      * Retries a specific request.
      *
      * @param request Request to retry.
      * @param context Context of the retry request.
+     * @param numberOfRetryAttempts Number of retry attempts.
      * @return {@code true} if a request can be retried; {@code false} otherwise;
      */
-    public abstract boolean doRetryRequest(REQ request, RequestContext context);
+    public abstract boolean doRetryRequest(REQ request, RequestContext context, int numberOfRetryAttempts);
+  }
+
+  @ThreadSafe
+  private static class ClientRetryTracker
+  {
+    private static final int COUNTER_TOTAL_COUNT_INDEX = 0;
+    private static final int COUNTER_RETRY_COUNT_INDEX = 1;
+
+    private final int _aggregatedIntervalNum;
+    private final long _updateIntervalMs;
+    private final Clock _clock;
+
+    private final Object _counterLock = new Object();
+    private final Object _updateLock = new Object();
+
+    @GuardedBy("_updateLock")
+    private volatile long _lastRollOverTime;
+    private double _currentAggregatedRetryRatio;
+
+    @GuardedBy("_counterLock")
+    private final LinkedList<int[]> _retryCounter;
+    private final int[] _aggregatedRetryCounter;
```

Comment:
I feel using int[2] to track total count and retry count is a bit implicit without java docs, can we use a new data structure to represent the count?

Reply:
Fixed. Added a RetryCounter impl.